tests/net/mod.rs

```diff
@@ -984,57 +984,40 @@ where
 
 impl<C, D, N> VirtualNet<D>
 where
-    D: DistAlgorithm<Output = Batch<C, N>>,
+    D: DistAlgorithm<NodeId = N, Output = Batch<C, N>>,
     C: Contribution + Clone,
     N: NodeIdT,
 {
-    /// Verifies that all nodes' outputs agree, and returns the output of node 1. Requirements
-    /// towards the output from node 0 are weaker compared to other nodes: it can be removed - hence
-    /// could miss some batches - but then it should rejoin the network.
-    ///
-    /// There must be at least nodes 0 and 1 in the network for this verification function to
-    /// complete.
-    pub fn verify_batches(&self) -> &[Batch<C, N>] {
-        const EXPECTED_ID: usize = 1;
-        let expected = self.correct_nodes().nth(EXPECTED_ID).unwrap().outputs();
-        let pub_eq = |(b0, b1): (&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
-        let pub_eq_ref = |(b0, b1): &(&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
-        for (i, node) in self
-            .correct_nodes()
-            .enumerate()
-            .filter(|(i, _)| *i != EXPECTED_ID)
-        {
-            if i == 0 {
-                assert!(expected.len() >= node.outputs().len());
-                assert!(
-                    expected
-                        .iter()
-                        .zip(node.outputs())
-                        .take_while(pub_eq_ref)
-                        .next()
-                        .is_some(),
-                    "Outputs of nodes 0 and {} don't start the same",
-                    EXPECTED_ID,
-                );
-                assert!(
-                    node.outputs()
-                        .iter()
-                        .find(|&b| !expected.iter().any(|c| b.public_eq(c)))
-                        .is_none(),
-                    "Node 0 has unexpected output"
-                );
-            } else {
-                assert!(
-                    expected.iter().zip(node.outputs()).all(pub_eq),
-                    "Outputs of nodes {} and {} differ: {:?} != {:?}",
-                    EXPECTED_ID,
-                    i,
-                    expected,
-                    node.outputs()
-                );
+    /// Verifies that all nodes' outputs agree, given a correct "full" node that received all
+    /// batches. Validity of other nodes' output is judged on the grounds of their contributions
```

Comment:
That's filtering out too many batches: A node may have been too slow, or it may have been an observer or a candidate; then it won't have a contribution in the batch, but is still expected to output it.

Reply:
I think the right thing to do is something like this:

```rust
let mut participants = unimplemented!("all initial nodes?");
let mut expected = BTreeMap::new();
for batch in &full_node.outputs {
    for id in participants {
        expected.entry(*id).or_default().push(batch);
    }
    unimplemented!("update participants according to batch.change()");
}
for node in self.correct_nodes().filter(|n| n.id() != full_node.id()) {
    assert_eq!(node.outputs, expected[node.id()]);
}
```

But in the long run I guess `VirtualNet` should support actually adding and removing nodes to/from the network? Not sure. Adding @mbr for ideas.